### PR TITLE
Merge blocks from peers with different or nil checksums during peer bootstrapping

### PR DESCRIPTION
**Pull request by xichen2020**
cc @robskillington @martin-mao

### CODE REVIEW

File: client/session.go [hunk 1]
```diff
@@ -1716,13 +1738,88 @@ func (r *blocksResult) addBlockFromPeer(id ts.ID, block *rpc.Block) error {
 	// No longer need the encoder, seal the block
 	result.Seal()
 
-	r.Lock()
-	r.result.AddBlock(id, result)
-	r.Unlock()
+	resultCtx := context.NewContext()
```
Comment (robskillington) [inline client/session.go]: Get this from a pool too?

File: client/session.go [hunk 2]
```diff
@@ -1716,13 +1738,88 @@ func (r *blocksResult) addBlockFromPeer(id ts.ID, block *rpc.Block) error {
...
+		result.Reset(start, encoder)
+		result.Seal()
+	}
 
 	return nil
 }
 
+func (r *blocksResult) mergeReaders(start time.Time, readers []io.Reader) (encoding.Encoder, error) {
+	alloc := r.opts.ReaderIteratorAllocate()
+	iter := encoding.NewMultiReaderIterator(alloc, nil)
```
Comment (robskillington) [inline client/session.go]: We should use the multi reader iterator pool to get this here I think?

File: client/session.go [hunk 3]
```diff
@@ -1716,13 +1738,88 @@ func (r *blocksResult) addBlockFromPeer(id ts.ID, block *rpc.Block) error {
...
+	iter.Reset(readers)
+	defer iter.Close()
+
+	encoder := r.encoderPool.Get()
+	encoder.Reset(start, r.blockAllocSize)
+
+	for iter.Next() {
+		dp, unit, annotation := iter.Current()
+		encoder.Encode(dp, unit, annotation)
```
Comment (robskillington) [inline client/session.go]: We should probably check the return of `Encode(...)` and return if err here?

File: integration/peers_bootstrap_merge_peer_blocks_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	if testing.Short() {
+		t.SkipNow() // Just skip if we're doing a short run
+	}
+
+	// Test setups
+	log := xlog.NewLogger(os.Stdout)
+	namesp := namespace.NewMetadata(testNamespaces[0], namespace.NewOptions())
+	opts := newTestOptions().
+		SetNamespaces([]namespace.Metadata{namesp})
+	testSleepFn, setSleepFn := newTestSleepFn()
```
Comment (xichen2020) [inline integration/peers_bootstrap_merge_peer_blocks_test.go]: sounds good

File: client/session.go [hunk 5]
```diff
@@ -1716,13 +1738,88 @@ func (r *blocksResult) addBlockFromPeer(id ts.ID, block *rpc.Block) error {
...
+
+		// Remove the existing block from the result so it doesn't get
+		// merged again
+		r.result.RemoveBlockAt(id, start)
+		r.Unlock()
+
+		// If we've already received data for this block, merge them
+		// with the new block if possible
+		if tmpCtx == nil {
+			tmpCtx = context.NewContext()
```
Comment (xichen2020) [inline client/session.go]: will do

File: client/session.go (hunk 1 above)
Comment (xichen2020) [inline client/session.go]: yeah

File: client/session.go (hunk 2 above)
Comment (xichen2020) [inline client/session.go]: yeah

File: client/session.go (hunk 3 above)
Comment (xichen2020) [inline client/session.go]: will do, took this from existing code so didn't pay too much attention there

File: client/session.go [hunk 6]
```diff
@@ -1403,79 +1404,113 @@ func (s *session) selectBlocksForSeriesFromPeerBlocksMetadata(
...
-					append(currStart[i].blocks[idx].reattempt.attempted, peer)
 				currStart[i].blocks[idx].reattempt.peersMetadata = peersMetadata
 
-				// Leave the block in the current peers blocks list
-				continue
+				// Each block now has attempted multiple peers
+				for j := range currStart {
+					currStart[i].blocks[idx].reattempt.attempt++
+					currStart[i].blocks[idx].reattempt.attempted =
+						append(currStart[i].blocks[idx].reattempt.attempted, currStart[j].peer)
```
Comment (robskillington) [inline client/session.go]: At some point we should rethink our retry semantics here.  If any of these fail they won't be retried against the peer we want them from. 

Let me open an issue about it.  For now this is fine, but we probably want real retry semantics in the future for trying to retrieve from the same peer on failure.

File: client/session.go [hunk 7]
```diff
@@ -1403,79 +1404,113 @@ func (s *session) selectBlocksForSeriesFromPeerBlocksMetadata(
...
-				currStart[i].blocks[idx].reattempt.attempt++
-				currStart[i].blocks[idx].reattempt.attempted =
-					append(currStart[i].blocks[idx].reattempt.attempted, peer)
 				currStart[i].blocks[idx].reattempt.peersMetadata = peersMetadata
 
-				// Leave the block in the current peers blocks list
-				continue
+				// Each block now has attempted multiple peers
+				for j := range currStart {
+					currStart[i].blocks[idx].reattempt.attempt++
```
Comment (robskillington) [inline client/session.go]: Need to set `reattempt.id` here too

File: client/session.go [hunk 8]
```diff
@@ -1403,79 +1404,113 @@ func (s *session) selectBlocksForSeriesFromPeerBlocksMetadata(
...
+			peersMetadata := make([]blockMetadataReattemptPeerMetadata, 0, len(currStart))
+			for i := range currStart {
+				unselected := currStart[i].unselectedBlocks()
+				metadata := blockMetadataReattemptPeerMetadata{
+					peer:     currStart[i].peer,
+					start:    unselected[0].start,
+					size:     unselected[0].size,
+					checksum: unselected[0].checksum,
+				}
+				peersMetadata = append(peersMetadata, metadata)
```
Comment (xichen2020) [inline client/session.go]: Sure thing

**Pull request merged**